cunumeric/array.py

```diff
@@ -2994,6 +2994,31 @@ def reshape(self, shape, order="C") -> ndarray:
         --------
         Multiple GPUs, Multiple CPUs
         """
+
+        computed_shape = (shape,) if isinstance(shape, int) else shape
+
+        if sum(extent == -1 for extent in computed_shape) > 1:
+            raise ValueError("can only specify one unknown dimension")
+
+        unknown_extent = self.size
+        for extent in computed_shape:
+            if extent == -1 or extent == 0:
+                continue
+            if unknown_extent % extent != 0:
+                raise ValueError(
+                    f"cannot reshape array of size {self.size} "
+                    f"into shape ({extent})"
+                )
+            unknown_extent //= extent
```

Comment:
I think if you add this check here:

```
if all(extent >= 0 for extent in computed_shape) and unknown_extent != 1:
  raise ValueError(f"cannot reshape array of shape {self.shape} into shape {computed_shape}")
```

then we can get rid of the `shape != -1` block below.

Reply:
This is still missing. We should have a check like this to catch this case early:

```
x = cn.ones((25,))
print(x.reshape(1, 5).shape)
```

currently this fails later on:

```
Traceback (most recent call last):
  File "/Users/mpapadakis/legate.core/install/lib/python3.8/site-packages/legion_top.py", line 440, in legion_python_main
    run_path(args[start], run_name='__main__')
  File "/Users/mpapadakis/legate.core/install/lib/python3.8/site-packages/legion_top.py", line 255, in run_path
    exec(code, module.__dict__, module.__dict__)
  File "a.py", line 13, in <module>
    print(x.reshape(1, 5).shape)
  File "/Users/mpapadakis/cunumeric/cunumeric/coverage.py", line 107, in wrapper
    return func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/array.py", line 3030, in reshape
    thunk=self._thunk.reshape(computed_shape, order),
  File "/Users/mpapadakis/cunumeric/cunumeric/deferred.py", line 961, in reshape
    out_prod *= out_shape[out_dim]
IndexError: tuple index out of range
```

Reply:
[f7475cd](https://github.com/nv-legate/cunumeric/pull/490/commits/f7475cd3365cec6a3edf28ae094103497a4993ea)